allow futures to be awaited and yielded from

lime-green opened the pull request:
Currently Pykka futures are not compatible with the `await` and `yield from` syntax. This PR simply adds that capability by implementing `__await__` and `__iter__` on the base Future class

lime-green commented:
Darn, `2.7` doesn't support returning inside a generator, and `3.7` doesn't support `StopIteration` inside one. I'll have to think about how to solve that

jodal commented on tests/conftest.py:
```diff
@@ -15,6 +17,20 @@
     ['name', 'actor_class', 'event_class', 'future_class', 'sleep_func'],
 )
 
+running_py3 = sys.version_info >= (3, 4, 0)
```
Use `pykka.compat.PY3` for this?

jodal commented on tests/test_future_py3.py:
```diff
@@ -0,0 +1,23 @@
+from tests import has_async_await
+
+def _run_async(coroutine):
+    import asyncio
```
Move to top of module, as it is only executed on Python 3?

jodal commented on tests/test_future_py3.py:
```diff
@@ -0,0 +1,23 @@
+@has_async_await
+def test_future_supports_await_syntax(future):
+    async def _get_value():
+        return await future
```
Please format with Black. I think it'll add a blank line after the inner function.

lime-green commented:
@jodal Ok I resolved all the issues 😅 Thanks for your quick review :)

lime-green commented:
@jodal anything I should do to get this merged?

jodal commented:
No, I don't think so. I've been away for a few days, so I haven't gotten around to this yet.

jodal commented on tests/test_future_py3.py:
```diff
@@ -0,0 +1,31 @@
+        return val
+
+    future.set(1)
+    assert _run_async(_get_value()) == 1
```
No need to underscore prefix these two functions as they are part of the test suite and not the module anyway, and thus will never be mistaken as a public API.

jodal commented on MANIFEST.in:
```diff
@@ -14,3 +14,5 @@ prune docs/_build
 
 recursive-include tests *.py
+
+recursive-include pykka *.py
```
I fixed this in `develop` now, so if you rebase, the change to `MANIFEST.in` in this branch should be redundant.

lime-green commented on dev-requirements.txt:
```diff
@@ -25,3 +25,6 @@ wheel
 twine
+
+# Python 3.4 doesn't have asyncio in the stdlib
+asyncio; python_version == '3.4'
```
Thanks! My mistake

lime-green commented:
@jodal I addressed all your comments. You should enable builds for forked pull requests:
> By default, CircleCI does not build PRs from forked repositories. To change this setting, go to the Advanced Settings of your project and set the Build forked pull requests option to On.
(https://circleci.com/docs/2.0/oss/)

jodal commented on tests/test_future_py3.py:
```diff
@@ -0,0 +1,31 @@
+    return loop.run_until_complete(f)
+
+
+@has_async_await
```
Looking at the end-of-life schedule, it seems Python 3.4 has 14 days left to live. I that this is now just 3 lines of code to remove when we drop Python 3.4 support.

jodal commented:
Thanks for the pointer! I've turned on builds of pull requests from forks now. CircleCI is new to me as of this week, but loving it so far :-)

lime-green commented:
@jodal Yeah! I’ve also used it a little bit and I was quite impressed with it as well :) Thanks for all your work on this library!
